go/arrow/avro/reader_types.go

```diff
@@ -0,0 +1,892 @@
+	switch dt := data.(type) {
+	case nil:
+		b.AppendNull()
+	case string:
+		b.AppendString(dt)
+	case map[string]any:
+		switch dt["string"].(type) {
+		case nil:
+			b.AppendNull()
+		case string:
+			b.AppendString(dt["string"].(string))
+		}
+	}
+}
+
+func appendBoolData(b *array.BooleanBuilder, data interface{}) {
+	switch dt := data.(type) {
+	case nil:
+		b.AppendNull()
+	case bool:
+		b.Append(dt)
+	case map[string]any:
+		switch dt["boolean"].(type) {
+		case nil:
+			b.AppendNull()
+		case bool:
+			b.Append(dt["boolean"].(bool))
+		}
+	}
+}
+
+func appendDate32Data(b *array.Date32Builder, data interface{}) {
+	switch dt := data.(type) {
+	case nil:
+		b.AppendNull()
+	case int32:
+		b.Append(arrow.Date32(dt))
+	case map[string]any:
+		switch dt["int"].(type) {
+		case nil:
+			b.AppendNull()
+		case int32:
+			b.Append(arrow.Date32(dt["int"].(int32)))
+		}
+	}
+}
+
+func appendDecimal128Data(b *array.Decimal128Builder, data interface{}) error {
+	switch dt := data.(type) {
+	case nil:
+		b.AppendNull()
+	case []byte:
+		if len(dt) <= 38 {
+			var intData int64
+			buf := bytes.NewBuffer(dt)
+			err := binary.Read(buf, binary.BigEndian, &intData)
+			if err != nil {
+				return err
+			}
+			b.Append(decimal128.FromI64(intData))
+		} else {
+			var bigIntData big.Int
+			buf := bytes.NewBuffer(dt)
+			err := binary.Read(buf, binary.BigEndian, &bigIntData)
+			if err != nil {
+				return err
+			}
+			b.Append(decimal128.FromBigInt(&bigIntData))
+		}
+	case map[string]any:
+		if len(dt["bytes"].([]byte)) <= 38 {
+			var intData int64
+			buf := bytes.NewBuffer(dt["bytes"].([]byte))
+			err := binary.Read(buf, binary.BigEndian, &intData)
+			if err != nil {
+				return err
+			}
+			b.Append(decimal128.FromI64(intData))
+		} else {
+			var bigIntData big.Int
+			buf := bytes.NewBuffer(dt["bytes"].([]byte))
+			err := binary.Read(buf, binary.BigEndian, &bigIntData)
+			if err != nil {
+				return err
+			}
+			b.Append(decimal128.FromBigInt(&bigIntData))
+		}
+	}
+	return nil
+}
+
+func appendDecimal256Data(b *array.Decimal256Builder, data interface{}) error {
+	switch dt := data.(type) {
+	case nil:
+		b.AppendNull()
+	case []byte:
+		var bigIntData big.Int
+		buf := bytes.NewBuffer(dt)
+		err := binary.Read(buf, binary.BigEndian, &bigIntData)
+		if err != nil {
+			return err
+		}
+		b.Append(decimal256.FromBigInt(&bigIntData))
+	case map[string]any:
+		var bigIntData big.Int
+		buf := bytes.NewBuffer(dt["bytes"].([]byte))
+		err := binary.Read(buf, binary.BigEndian, &bigIntData)
+		if err != nil {
+			return err
+		}
+		b.Append(decimal256.FromBigInt(&bigIntData))
+	}
+	return nil
+}
+
+func appendDurationData(b *array.MonthDayNanoIntervalBuilder, data interface{}) {
+	switch dt := data.(type) {
+	case nil:
+		b.AppendNull()
+	case []byte:
+		dur := new(arrow.MonthDayNanoInterval)
+		dur.Months = int32(binary.BigEndian.Uint64(dt[:3]))
+		dur.Days = int32(binary.BigEndian.Uint64(dt[4:7]))
```

Comment:
I'm confused by these indices, Avro stores them as only 3 bytes for each value? not 4 bytes? and why not use `.Uint32` instead of `Uint64` ?

Reply:
my mistake, incorrect use of the high bound in slicing, I thought it was inclusive whereas the high bound is exclusive.
I've corrected it to :
```
dur.Months = int32(binary.LittleEndian.Uint16(dtb[:4]))
dur.Days = int32(binary.LittleEndian.Uint16(dtb[4:8]))
dur.Nanoseconds = int64(binary.LittleEndian.Uint32(dtb[8:]) * 1000000)
```